Dapper/SqlMapper.Settings.cs

```diff
@@ -7,6 +10,47 @@ partial class SqlMapper
         /// </summary>
         public static class Settings
         {
+            // disable single result by default; prevents errors AFTER the select being detected properly
+            const CommandBehavior DefaultAllowedCommandBehaviors = ~CommandBehavior.SingleResult;
+            internal static CommandBehavior AllowedCommandBehaviors { get; private set; } = DefaultAllowedCommandBehaviors;
+            private static void SetAllowedCommandBehaviors(CommandBehavior behavior, bool enabled)
+            {
+                if (enabled) AllowedCommandBehaviors |= behavior;
+                else AllowedCommandBehaviors &= ~behavior;
+            }
+            /// <summary>
+            /// Gets or sets whether dapper should use the CommandBehavior.SingleResult optimization
+            /// </summary>
+            /// <remarks>Note that a consequence of enabling this option is that errors that happen <b>after</b> the first select may not be reported</remarks>
+            public static bool UseSingleResultOptimization
+            {
+                get { return (AllowedCommandBehaviors & CommandBehavior.SingleResult) != 0; }
+                set { SetAllowedCommandBehaviors(CommandBehavior.SingleResult, value); }
+            }
+            /// <summary>
+            /// Gets or sets whether dapper should use the CommandBehavior.SingleRow optimization
+            /// </summary>
+            /// <remarks>Note that on some DB providers this optimization can have adverse performance impact</remarks>
+            public static bool UseSingleRowOptimization
+            {
+                get { return (AllowedCommandBehaviors & CommandBehavior.SingleRow) != 0; }
+                set { SetAllowedCommandBehaviors(CommandBehavior.SingleRow, value); }
+            }
+            internal static bool DisableCommandBehaviorOptimizations(CommandBehavior behavior, Exception ex)
+            {
+                if (AllowedCommandBehaviors == DefaultAllowedCommandBehaviors
+                    && (behavior & (CommandBehavior.SingleResult | CommandBehavior.SingleRow)) != 0)
+                {
+                    if (ex.Message.Contains(nameof(CommandBehavior.SingleResult))
+                        || ex.Message.Contains(nameof(CommandBehavior.SingleRow)))
+                    { // some providers just just allow these, so: try again without them and stop issuing them
+                        SetAllowedCommandBehaviors(CommandBehavior.SingleResult | CommandBehavior.SingleRow, false);
+                        return true;
+                    }

```

Comment:
IMO, these should be independent disables - would split this if/disable into one per `CommandBehavior`. For example if we added a third `CommandBehavior` that auto-disables this would get hairy...baby/bathwater situation possibly. I'm trying to think of a better detection method than this but I don't have better suggestions, +1 on going with the `.Contains()` you have.